src/libopensc/card-piv.c

```diff
@@ -2283,31 +2288,42 @@ static int piv_validate_general_authentication(sc_card_t *card,
 					u8 * out, size_t outlen)
 {
 	piv_private_data_t * priv = PIV_DATA(card);
-	int r;
+	int r, tmplen, tmplen2;
 	u8 *p;
 	const u8 *tag;
 	size_t taglen;
 	const u8 *body;
 	size_t bodylen;
-	unsigned int real_alg_id;
+	unsigned int real_alg_id, op_tag;
 
 	u8 sbuf[4096]; /* needs work. for 3072 keys, needs 384+10 or so */
+	size_t sbuflen = sizeof(sbuf);
 	u8 rbuf[4096];
 
 	SC_FUNC_CALLED(card->ctx, SC_LOG_DEBUG_VERBOSE);
 
 	/* should assume large send data */
 	p = sbuf;
-	put_tag_and_len(0x7c, (2 + put_tag_and_len(0, datalen, NULL)) , &p);
-	put_tag_and_len(0x82, 0, &p);
+	tmplen = sc_asn1_put_tag(0xff, NULL, datalen, NULL, 0, NULL);
+	tmplen2 = sc_asn1_put_tag(0x82, NULL, 0, NULL, 0, NULL);
+	if (tmplen <= 0 || tmplen2 <= 0) {
+		LOG_FUNC_RETURN(card->ctx, SC_ERROR_INTERNAL);
+	}
+	tmplen += tmplen2;
+	if ((r = sc_asn1_put_tag(0x7c, NULL, tmplen, p, sbuflen, &p)) != SC_SUCCESS ||
+	    (r = sc_asn1_put_tag(0x82, NULL, 0, p, sbuflen - (p - sbuf), &p)) != SC_SUCCESS) {
+		LOG_FUNC_RETURN(card->ctx, r);
+	}
 	if (priv->operation == SC_SEC_OPERATION_DERIVE
-			&& priv->algorithm == SC_ALGORITHM_EC)
-		put_tag_and_len(0x85, datalen, &p);
-	else
-		put_tag_and_len(0x81, datalen, &p);
-
-	memcpy(p, data, datalen);
-	p += datalen;
+			&& priv->algorithm == SC_ALGORITHM_EC) {
```

Comment:
clang-format suggested the following change:


```
-	if (priv->operation == SC_SEC_OPERATION_DERIVE
-			&& priv->algorithm == SC_ALGORITHM_EC) {
+	if (priv->operation == SC_SEC_OPERATION_DERIVE && priv->algorithm == SC_ALGORITHM_EC) {
```

<!-- identifier: clang-format -->
<!-- File: src/libopensc/card-piv.c -->